src/libostree/ostree-sysroot.c

```diff
@@ -374,11 +401,75 @@ remount_writable (const char *path, gboolean *did_remount, GError **error)
   return TRUE;
 }
 
+static gboolean
+is_sysroot_invisible (void)
+{
+  struct stat stbuf;
+
+  if (lstat (OTCORE_RUN_OSTREE_PRIVATE "/sysroot-ns", &stbuf) < 0)
+    return FALSE;
+
+  if (lstat ("/sysroot/ostree", &stbuf) == 0)
+    return FALSE;
+
+  return TRUE;
+}
+
+/* Unmount covering tmpfs to make /sysroot visible */
+gboolean
+_ostree_sysroot_ensure_visible (OstreeSysroot *self, GError **error)
+{
+  if (!ostree_sysroot_initialize (self, error))
+    return FALSE;
+
+  /* Do nothing if no mount namespace is in use */
+  if (!self->mount_namespace_in_use)
+    return TRUE;
+
+  /* If we aren't operating on a booted system, then we don't
+   * do anything with mounts.
+   */
+  if (!self->root_is_ostree_booted)
+    return TRUE;
+
+  /* Handle invisible sysroot */
+  if (is_sysroot_invisible ())
+    {
+      glnx_autofd int sysroot_ns_fd = open (OTCORE_RUN_OSTREE_PRIVATE "/sysroot-ns", O_RDONLY);
+      if (sysroot_ns_fd < 0)
+        return FALSE;
+
+      glnx_autofd int cur_ns_fd = open("/proc/self/ns/mnt", O_RDONLY);
+      if (cur_ns_fd < 0)
+        return FALSE;
+
+      if (setns (sysroot_ns_fd, CLONE_NEWNS) < 0)
+        return FALSE;
+
+      glnx_autofd int tree_fd = open_tree (AT_FDCWD, "/", OPEN_TREE_CLONE);
+      if (tree_fd < 0)
+        return FALSE;
+
+      if (setns (cur_ns_fd, CLONE_NEWNS) < 0)
+        abort (); // it's unsafe to continue if we cannot switch back
+
+      if (move_mount (tree_fd, "", AT_FDCWD, "/sysroot", MOVE_MOUNT_F_EMPTY_PATH) < 0)
```

Comment:
Ideally we can drop this...everything in libostree and its callers should hopefully only access things using file-descriptors. There's some legacy tech debt in sysroot because the openat() stuff was only added later (and there's some GFile usage which was a mistake), but...maybe what we can do is just fail if anything tries to access the path property when we're in this mode.

Reply:
Sure. We can grab the fd using open_tree directly and do not need to enter a mount namespace. However, I'm afraid the caller may still use paths somewhere and it will break the compatibility.

Reply:
I made it in 96e2fb8. PTAL.